### PR TITLE
Support additional Coveo custom tags

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
### Generate Tags from Build Option Arguments

The following build option arguments are now leveraged as custom tags in the generated reference documentation:

- defaultValue
- depend
- min
- max
- float
- required

##### min, max, defaultValue

Screen Shot 2019-09-19 at 12 11 35 PM
Screen Shot 2019-09-19 at 11 56 09 AM

##### float

Screen Shot 2019-09-19 at 12 30 07 PM
Screen Shot 2019-09-19 at 12 34 30 PM

##### depend

Screen Shot 2019-09-19 at 12 21 24 PM
Screen Shot 2019-09-19 at 12 21 38 PM

##### required

Screen Shot 2019-09-19 at 12 25 27 PM
Screen Shot 2019-09-19 at 12 26 58 PM

### Support New Custom Tags

In addition to @notsupportedin and @examples, the following tags are now supported:

- @availablesince
- @deprecatedsince
- @externaldocs
- @isresulttemplatecomponent

Screen Shot 2019-09-19 at 12 46 49 PM
Screen Shot 2019-09-19 at 12 46 40 PM

### Code Refactoring

I started moving Coveo-specific logic into a distinct file / namespace. This is obviously a work in progress, but I think it's going to make it easier for us if we want to pull the latest changes from the main TypeDoc repository, for instance.

### CODE REVIEW

**Comment by fbeaudoincoveo** [comment]
Sorry guys, this is quite a big PR :/

File: src/lib/coveo/coveoCustom.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
+import { Reflection, ReflectionFlag, CommentTag } from "../models";
+import { parseComment } from "../converter/factories/comment";
+import { Converter } from "../converter";
+
+export interface ICoveoDisplayTemplateFunction {
+  (rawValue?: string, result?: Reflection): string;
+}
+
+export class CoveoCustomTag {
```
Comment (samisayegh) [inline src/lib/coveo/coveoCustom.ts:9]: I think this could become a function rather than a class,

File: src/lib/coveo/coveoCustom.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+export interface IParsedCoveoBuildOptionArgs {
+  required?: CoveoCustomTag,
+  defaultValue?: CoveoCustomTag,
+  depend?: CoveoCustomTag,
+  min?: CoveoCustomTag,
+  max?: CoveoCustomTag,
+  float?: CoveoCustomTag
+}
+
+export class CoveoCustom {
```
Comment (samisayegh) [inline src/lib/coveo/coveoCustom.ts:32]: There's quite a bit of logic here, I wonder if we should start adding unit tests to this project 😬

**Review by samisayegh — APPROVED** [review]

File: src/lib/coveo/coveoCustom.ts (hunk 2 above)
Comment (fbeaudoincoveo) [reply src/lib/coveo/coveoCustom.ts:32]: It would probably be a good idea, yes :)

File: src/lib/converter/converter.ts [hunk 3]
```diff
@@ -310,49 +310,20 @@ export class Converter extends ChildableComponent<Application, ConverterComponen
         context.visitStack = oldVisitStack.slice();
         context.visitStack.push(node);
 
-        let result: Reflection;
+        let result: Reflection
```
Comment (ThibodeauJF) [inline src/lib/converter/converter.ts]: Why remove the `;`?

Comment (fbeaudoincoveo) [reply src/lib/converter/converter.ts]: Seems like an accident, thanks for pointing this out!

File: src/lib/coveo/coveoCustom.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    return parsedBuildOptionArgs;
+  }
+
+  static areResultAndCommentDefined(result: Reflection, comment: string|null) {
+    return result && comment !== null;
+  }
+
+  static setNotSupportedInTag(converter: Converter, result: Reflection, comment: string) {
+    if (comment.indexOf('@notSupportedIn') !== -1) {
+      var tagRegex = /@(?:notSupportedIn)\s*((?:[\w]+, )*[\w]+)/g;
```
Comment (ThibodeauJF) [inline src/lib/coveo/coveoCustom.ts]: use const/let instead of var

**Review by ThibodeauJF — APPROVED** [review]

**Pull request merged**